### PR TITLE
Add type and format to property definition

### PR DESCRIPTION
**Pull request by pauldambra**
## Changes

For ##6619 and following on from (but not depending on) #7608 

Adds a migration to add two new fields to property definitions

### property_type field

a restricted set of options of what type a field is
initially only allowing DateTime

### property_type_format field

an optional format for the property type. E.g; for a date selecting between YYYY-MM-DD or a UNIX timestamp

## How did you test this code?

by running the API locally and by writing tests

### CODE REVIEW

File: posthog/migrations/0194_set_property_type_for_time.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    set property_type = 'DateTime', property_type_format='unix_timestamp'
+    where name = '$time'
+        """,
+            reverse_sql="""
+                update posthog_propertydefinition
+    set property_type = null, property_type_format=null
+    where name = '$time'
+                """,
+        )
+    ]
```
Comment (mariusandra) [inline posthog/migrations/0194_set_property_type_for_time.py:25]: This migration will now run only once... but what about for users who install a fresh posthog instance, run all the migrations as needed, and only then ingest their first event which contains this property?

File: plugin-server/tests/shared/process-event.ts [hunk 2]
```diff
@@ -136,7 +136,7 @@ export const createProcessEventTests = (
     async function processEvent(
         distinctId: string,
         ip: string | null,
-        siteUrl: string,
+        _siteUrl: string,
```
Comment (mariusandra) [inline plugin-server/tests/shared/process-event.ts:139]: This looks like an interesting flyby change?

File: posthog/models/property_definition.py [hunk 3]
```diff
@@ -5,6 +5,16 @@
 from posthog.models.utils import UUIDModel
 
 
+class PropertyType(models.TextChoices):
+    Datetime = "DateTime", "DateTime"
```
Comment (mariusandra) [inline posthog/models/property_definition.py:9]: Should we already add more values here? Adding enums later in django requires another db migration to be made and run.

**Review by mariusandra — APPROVED** [review]
Basically looks good. Some comments inline, but if you have a plan for them, I don't see any reason to hold this up

File: posthog/migrations/0194_set_property_type_for_time.py (hunk 1 above)
Comment (pauldambra) [reply posthog/migrations/0194_set_property_type_for_time.py:25]: Yep, going to add smarts later on to do detection and allow editing

File: plugin-server/tests/shared/process-event.ts (hunk 2 above)
Comment (pauldambra) [reply plugin-server/tests/shared/process-event.ts:139]: I was obeying the linter. Unused parameter

File: posthog/models/property_definition.py (hunk 3 above)
Comment (pauldambra) [reply posthog/models/property_definition.py:9]: I was wondering about this... I was anticipating the migration would be cheap to run (drop and add a constraint) Am expecting that I'll add more before proper release. Wanted to get DateTime in so I can prove the shape without thinking about too much at once

Comment (mariusandra) [reply posthog/models/property_definition.py:9]: I guess it'll be fine. The table we're operating on won't have *that* many rows in the end (few million max?), so even if it's a more expensive index query (doubtful in the case of a constraint?), I think we're fine. Though I've been burned by postgres migrations more times than I can count, so take from this what you want :D.

Comment (pauldambra) [reply posthog/models/property_definition.py:9]: Yeah, no harm in adding string, number, bigint, and boolean to match https://github.com/PostHog/posthog/blob/master/frontend/src/lib/components/PropertiesTable.tsx#L12

As things stand we'd need a migration each time we add a format to update the check constraint

Comment (pauldambra) [reply posthog/models/property_definition.py:9]: @mariusandra added the types to bring the format in-line with the types that the `PropertiesTable` knows how to display

**Review by mariusandra — APPROVED** [review]

**Pull request merged**